webrender_traits/src/types.rs

```diff
@@ -55,7 +55,25 @@ pub enum ApiMsg {
     WebGLCommand(WebGLContextId, WebGLCommand),
     GenerateFrame,
     // WebVR commands that must be called in the WebGL render thread.
-    VRCompositorCommand(WebGLContextId, VRCompositorCommand)
+    VRCompositorCommand(WebGLContextId, VRCompositorCommand),
+    // An opaque handle that must be passed to the render notifier. It is used by Gecko
+    // to forward gecko-specific messages to the render thread preserving the ordering
+    // within the other messages.
+    ExternalEvent(ExternalEvent),
+}
+
+#[derive(Clone, Deserialize, Serialize)]
+pub struct ExternalEvent {
+    // An opaque pointer-sized value.
+    raw: usize,
+}
+
+unsafe impl Send for ExternalEvent {}
+
+impl ExternalEvent {
+    pub fn new(raw: usize) -> Self { ExternalEvent { raw: raw } }
+    // Consume self to make it obvious that the callback should be forwarded only once.
```

Comment:
Clone is derived because ApiMsg derives Clone which apparently needs to implement it for RenderApiSender to implement Clone as well. I think that it boils down to MsgSender<T> using derive(Clone) instead of implementing it manually, forcing T to implement it as well even though cloning the sender does not clone any message. I guess I should entangle that.